src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsReadStream.cs

```diff
@@ -0,0 +1,350 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Data.SqlClient;
+
+namespace Microsoft.Data.SqlClientX.IO
+{
+    /// <summary>
+    /// A stream which handles reading TDS messages from the wire, and extracting the data from the 
+    /// packets in the message packets.
+    /// </summary>
+    internal class TdsReadStream : Stream, ITdsReadStream
+    {
+        #region Private Fields
+        
+        private Stream _underlyingStream;
+
+        // The buffer to hold the TDS read data.
+        private byte[] _readBuffer;
+
+        // The read pointer inside the buffer.
+        private int _readIndex { get; set; } = 0;
+
+        // The end of the data index in the buffer.
+        private int _readBufferDataEnd { get; set; } = 0;
+
+        /// <summary>
+        /// The number of bytes left in the packet to be consumed.
+        /// This is not necessarily the number of bytes available in the buffer.
+        /// </summary>
+        private int _packetDataLeft { get; set; } = 0;
+
+        /// <summary>
+        /// The number of bytes in the header, according to the header.
+        /// </summary>
+        private int _packetHeaderDataLength { get; set; } = 0;
+
+        private byte _packetHeaderType { get; set; } = 0;
+
+        private byte _packetStatus { get; set; }
+
+        /// <summary>
+        /// Stored internally. It will be used for tracing.
+        /// </summary>
+        private int _spid;
```

Comment:
Have you given a thought about how would this be exposed via the public API [SqlConnection.ServerProcessId](https://learn.microsoft.com/en-us/dotnet/api/microsoft.data.sqlclient.sqlconnection.serverprocessid?view=sqlclient-dotnet-standard-5.2)?

Reply:
I hadnt realized that this field has a public surface area associated with it. 

I will make this a Property which will be gettable and defaults to 0 (to match the current SqlConnection behavior). 
Then this needs to be exposed via `TdsStream` 

The SqlConnection will use the SqlInternalConnection, and that will have a handle to TdsStream, which can then expose the SPID on the internal connection via TdsStream.